### ISSUE TITLE
Inconsistent authorization between az_rm_dnsrecordset_facts and az_rm_dnsrecordset

### ISSUE THREAD

**Issue by jdashton** [Azure/azure_preview_modules#279]
When I perform this action, the credentials in use have sufficient authorization:
```
  module_defaults:
    group/azure:
      resource_group: nad-core
      subscription_id: "{{ azuredns_subscriptionid }}"
      client_id: "{{ azuredns_appid }}"
      secret: "{{ azuredns_clientsecret }}"
      tenant: "{{ azuredns_tenantid }}"

  tasks:
    - name: Get Azure DNS Zone facts
      azure_rm_dnsrecordset_facts:
        zone_name: hope4humanity.info
        relative_name: 'jda'
        record_type: A
```
But when I perform this action with the same credentials on the same resource_group, zone_name, relative_name and record_type, I get an error about not having "read" authorization on the resource_group:
```
  module_defaults:
    group/azure:
      resource_group: nad-core
      subscription_id: "{{ azuredns_subscriptionid }}"
      client_id: "{{ azuredns_appid }}"
      secret: "{{ azuredns_clientsecret }}"
      tenant: "{{ azuredns_tenantid }}"

  tasks:
    - name: Get Azure DNS Zone facts
      azure_rm_dnsrecordset:
        zone_name: hope4humanity.info
        relative_name: 'jda'
        record_type: A
        records:
          - entry: 192.168.100.101
```
The service principal has limited authorization to do anything and everything with four specific domains out of a larger resource group. Clearly, from the first example above, it has sufficient authorization to read records. Why does this error message about "read" authorization come up when using the same credentials?
```
"msg": "Error retrieving resource group nad-core - The client '810b4e57-8616-4e20-bf8f-de89127fa05d' with object id '810b4e57-8616-4e20-bf8f-de89127fa05d' does not have authorization to perform action 'Microsoft.Resources/subscriptions/resourcegroups/read' over scope '/subscriptions/3f4c0514-e03d-4bdd-9572-c7b33fd31905/resourcegroups/nad-core'."
```
The service principal was created with this command:
```
az ad sp create-for-rbac --name  "AcmeDnsValidator" --role "DNS Zone Contributor" --scopes scope1 scope2 etc
```
The scopes in this command pointed to the four specific DNS zones within the larger resource group.

Also, the same credentials are sufficient, when used with [acme.sh](https://github.com/Neilpang/acme.sh), to create DNS records in these zones.

**Comment by jdashton** [comment]
In summary:

`az_rm_dnsrecordset.py` appears to be using a common validation function that assumes too wide an authorization for the specific work requested.

Here is a trace of the interesting bits from Ansible:
```
Using module file /usr/local/Cellar/ansible/2.8.1/libexec/lib/python3.7/site-packages/ansible/modules/cloud/azure/azure_rm_dnsrecordset.py
Pipelining is enabled.
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: jdashton
<localhost> EXEC /bin/sh -c '/usr/local/Cellar/ansible/2.8.1/libexec/bin/python3.7 && sleep 0'
The full traceback is:
  File "/var/folders/6v/szbw5tt53453v3dg49_lwl980000gn/T/ansible_azure_rm_dnsrecordset_payload__6ekwxnm/ansible_azure_rm_dnsrecordset_payload.zip/ansible/module_utils/azure_rm_common.py", line 449, in get_resource_group
    return self.rm_client.resource_groups.get(resource_group)
  File "/usr/local/Cellar/ansible/2.8.1/libexec/lib/python3.7/site-packages/azure/mgmt/resource/resources/v2017_05_10/operations/resource_groups_operations.py", line 293, in get
    raise exp

fatal: [aca-web-blue -> localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "ad_user": null,
            "adfs_authority_url": null,
            "api_profile": "latest",
            "append_tags": true,
            "auth_source": null,
            "cert_validation_mode": null,
            "client_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "cloud_environment": "AzureCloud",
            "password": null,
            "profile": null,
            "record_mode": "purge",
            "record_type": "A",
            "records": [
                {
                    "entry": "192.168.100.101",
                    "ipv4_address": "192.168.100.101"
                }
            ],
            "relative_name": "jda",
            "resource_group": "nad-core",
            "secret": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "state": "present",
            "subscription_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "tags": null,
            "tenant": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "time_to_live": 3600,
            "zone_name": "hope4humanity.info"
        }
    },
    "msg": "Error retrieving resource group nad-core - The client '810b4e57-8616-4e20-bf8f-de89127fa05d' with object id '810b4e57-8616-4e20-bf8f-de89127fa05d' does not have authorization to perform action 'Microsoft.Resources/subscriptions/resourcegroups/read' over scope '/subscriptions/3f4c0514-e03d-4bdd-9572-c7b33fd31905/resourcegroups/nad-core'."
}
```

**Comment by MyronFanQiu** [comment]
@jdashton Hello. Thanks for opening this issue! We will look into this issue.

**Comment by MyronFanQiu** [comment]
@jdashton Hi. I did some tests and checked the codes. `az_rm_dnsrecordset` and `az_rm_dnsrecordset_facts` do use same authorization method to get the `dnsrecordset` instance. From the logs, it seems that the `resource_group: nad-core` is created under other subscription or created by other tenant ID. I want to check whether the `resource_group: nad-core` is created by the same tenant id or not.

**Comment by jdashton** [comment]
Thanks for raising these issues, @MyronFanQiu. As a lowly developer, I don't have access to the entire `nad-core` resource group, and in the Azure portal I am permitted to see only the subscription and the four domains for which I am an owner. I can confirm that the subscription ID I see in the Azure portal or app matches the subscription ID being used in the credentials above.

I don't have a full grasp of the meaning and use of the tenant ID, but based on my limited research it appears that the `resource_group: nad-core` and the service principal had to be created by the same tenant id.

I can sign into the `az` CLI using my AD user ID and password. My AD user has the `owner` role for the four domains in question, which are just four of many domains in the `resource_group: nad-core`. Having signed-in, I used the commands listed on the `acme.sh` page linked above to create a service principal, assigning it `create-for-rbac` rights to each of the four domains for which my AD user is an owner.

I have limited access to the Azure controls for our subscription, but if you would like to recommend some `az` CLI commands that would show us whether these have the same tenant ID or not, I would be happy to run those commands and report the answers here. Thanks again for taking a look at this situation with us.

**Comment by jdashton** [comment]
Around like 299 of `azure_rm_dnsrecordset.py` are these two lines:
```
        # retrieve resource group to make sure it exists
        self.get_resource_group(self.resource_group)
```
These cause a problem if the service principal does not have permission to read the resource group, as in my case. If I comment out this call to `get_resource_group()`, the attempted DNS recordset update succeeds.

Commenting out this function call leaves us, of course, without a helpful error message if the specified resource group doesn't exist. If that happens, we'll see a somewhat misleading error message of the form 
```
Message: The client 'c' with object id 'c' does not have authorization to perform action 'Microsoft.Network/dnsZones/read' over scope '/subscriptions/sub1/resourceGroups/invalid-rg1/providers/Microsoft.Network/dnsZones/zone1'.
```
This message is technically accurate, but doesn't lead the programmer to realize that it's the resource group that is incorrect.

In `resource_groups_operations.py` around line 123 is a function definition for `check_existence()`. I haven't worked out yet whether this could be called instead of the `get_resource_group()`, and whether it would be more forgiving when the service principal does not have read permissions on the resource group.

**Comment by jdashton** [comment]
I tried a solution like this in place of the code snippet above (around line 299 of `azure_rm_dnsrecordset.py`):
```
        exists = self.rm_client.resource_groups.check_existence(self.resource_group)
        if not exists:
            self.fail('The resource group {0} does not exist'.format(self.resource_group))
```
I didn't know how to pass credentials, and so this function got an HTTP 403 error no matter whether the resource group exists or does not exist. 
```
msrestazure.azure_exceptions.CloudError: 403 Client Error: Forbidden for url: https://management.azure.com/subscriptions/sub1/resourcegroups/good-or-bad-rg1?api-version=2017-05-10
```

**Comment by MyronFanQiu** [comment]
@jdashton Good job. 👍 I just checked the python sdk and I think your solution is a possible and acceptable solution for this situation. When you call the `self.rm_client`, the credentials should have been already passed to the Azure. According to the log, in my understanding, the url should be like 
```
https://management.azure.com/subscriptions/xxxxxxxx-xxxx-xxxx-xxxxxxxxxxxx(subscription_id)/resourcegroups/myResourceGroup(resource_group_name)?api-version=2017-05-10
```
However, there is a word `sub1` in the url. Could you please paste the full stack trace log here?
I'm truly sorry it's not easy for me to setup the same situation and environment as yours and reproduce the same results. We need your help to solve this issue. 😄

**Comment by MyronFanQiu** [comment]
The url used in python sdk is 
```
    check_existence.metadata = {'url': '/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}'}
```
The credentials are passed through this call
https://github.com/Azure/azure_preview_modules/blob/d14822fc237a1309b156084e447ad25d404e5284/module_utils/azure_rm_common.py#L879-L885
In the `self.get_mgmt_svc_client`, the credentials will be passed.
https://github.com/Azure/azure_preview_modules/blob/d14822fc237a1309b156084e447ad25d404e5284/module_utils/azure_rm_common.py#L761

**Comment by jdashton** [comment]
Thanks for your help and encouragement, @MyronFanQiu.

I moved the call to `check_existence()` so that it is located in `azure_rm_common.py`, parallel to the `get()` call we had been using, so it should have the same set-up and credentials. It still returns a 403 error for both a real `resource_group` and for a non-existent `resource_group`'. I assume that my AD user and the service principal I created are forbidden from directly querying the existence of a resource group in this subscription.

In the following error messages, the first error message is for a non-existent resource group, `n4d-core`, and the second error message for the resource group that exists in this subscription, `nad-core`. (I'm replacing the actual subscription ID with `subid` in these error messages, as a nod to security.
```
Error retrieving resource group n4d-core - Operation failed with status: 'Forbidden'. Details: 403 Client Error: Forbidden for url: https://management.azure.com/subscriptions/subid/resourcegroups/n4d-core?api-version=2017-05-10
Error retrieving resource group nad-core - Operation failed with status: 'Forbidden'. Details: 403 Client Error: Forbidden for url: https://management.azure.com/subscriptions/subid/resourcegroups/nad-core?api-version=2017-05-10
```
The Azure adapter for `acme.sh` avoids this issue by calling `/subscriptions/subid/providers/Microsoft.Network/dnszones?api-version=2017-09-01` ([Zones - List](https://docs.microsoft.com/en-us/rest/api/dns/zones/list))to get a list of all DNS zones that can be seen. It then uses the zone ID of the selected domain from the results of that call. The zone ID already contains a resource group, which can be assumed to be valid and to exist. Here's the example zone ID from the above documentation page:
```
/subscriptions/subid/resourceGroups/rg1/providers/Microsoft.Network/dnsZones/zone1
```
In fixing the bug that I'm reporting on this ticket, I'm not sure whether to favor attempting to request the available DNS zones, as `acme.sh` does, as a means of validating that the zone and resource group both exist, or to take it on faith that the resource group provided by the caller must exist, but might be inaccessible to the caller, and then expand the error message to help guide the caller if the resource group does not exist.

**Comment by jdashton** [comment]
The error message you get when passing a non-existent resource group contains this excerpt:
```
does not have authorization to perform action 'Microsoft.Network/dnsZones/read' over scope
```
I would want to expand this message to suggest that the resource group might be invalid. However, the message is being generated at least two layers deeper into the library, and is being raised in `zones_operations.py` on line 268. It looks like it would be difficult to augment this message.

**Comment by jdashton** [comment]
Is there a place where I could create a temporary subscription, etc, in order to demonstrate this issue?

**Comment by MyronFanQiu** [comment]
@jdashton Yeah. I see. Let me clarify this issue for us and you can correct me if I misunderstand something.
There are two issues now.
I. For your specific situation, you don't have the authorization to read the resource_group. The existence check for the resource group will block you from updating DNS recordset. Thus you comment out the line 299 of `azure_rm_dnsrecordset.py`. It can temporarily unblock you from updating DNS recordset.
```
        # retrieve resource group to make sure it exists
        self.get_resource_group(self.resource_group)
```
II. After commenting out the existence check for resource group, if the resource_group doesn't exist, the error message isn't helpful. It will show the same error message for both situations where the resource group name is invalid or authorization fails. So you want to expand the message to suggest that the resource group might be invalid.
Do I miss something? Feel free to correct me.

**Comment by jdashton** [comment]
@MyronFanQiu Yes, you understand correctly. I would not have used the word "temporarily", because I'm hoping that you really remove those two lines. ;-) 

So, the new title for this issue would be **`azure_rm_dnsrecordset.py` assumes unnecessary access authorization** or something like that. The issue is that you don't need read permissions on the resource group to actually change the DNS record set contents, but this code assumes that you have read permissions, and fails unnecessarily if you don't.

I would be happy to submit a PR for this, and will do so if I can get to it quickly enough.

For item II, you are correct again. This isn't a problem for my specific situation, but I recognize that this (the lack of a helpful error message) might be grounds for rejecting the PR above. I hope to have time to try the technique of requesting to list of DNS zones and using that to verify that the provided resource group is probably valid for this request.

**Comment by jdashton** [comment]
Another approach I can try would be to just catch the exception being thrown by those two lines, and parse the exception to figure out (or guess) whether the caller is in this specific situation (the resource group exists, but you're not allowed to read it).

**Comment by MyronFanQiu** [comment]
@jdashton Yeah. It's cool about the PR. Thanks very much. I used the word 'temporarily' because we should find a way to solve your problem permanently. 😄 Your PR is one option. I agree that "azure_rm_dnsrecordset.py assumes unnecessary access authorization" is a reason for the PR. Of course, your another approach looks better. I will check whether it is a possible and easy option today. Maybe a few hours later and I will update the status here. Thanks you! 👍

**Comment by MyronFanQiu** [comment]
It seems that the `get_resource_group` can throw two kinds of error message. 
```
"msg": "Error retrieving resource group fanqiu-dontexist- Resource group 'fanqiu-dontexist' could not be found."
```
Maybe you can wrap the function `get_resource_group` by a new function, catch the exception and parse it to guess in which situation it is now.
Are you willing and having time to improve your PR? Or I can continue the work based on our discussion and your existed PR. 😄 Thanks~

**Comment by jdashton** [comment]
I'd like to submit a PR doing what you described. I won't be able to do it until Sunday at the earliest, but if you don't have the problem solved, I hope to submit a PR for it at that time.

**Comment by jdashton** [comment]
Hi again, @MyronFanQiu. The error messages seem to be indistinguishable when the credentials presented don't have read access at a level that should permit read access. Here are the two error messages, with the subscription ID obfuscated:
```
Error retrieving resource group jda-c0r3 - The client '810b4e57-8616-4e20-bf8f-de89127fa05d' with object id '810b4e57-8616-4e20-bf8f-de89127fa05d' does not have authorization to perform action 'Microsoft.Resources/subscriptions/resourcegroups/read' over scope '/subscriptions/subid/resourcegroups/jda-c0r3'.
Error retrieving resource group nad-core - The client '810b4e57-8616-4e20-bf8f-de89127fa05d' with object id '810b4e57-8616-4e20-bf8f-de89127fa05d' does not have authorization to perform action 'Microsoft.Resources/subscriptions/resourcegroups/read' over scope '/subscriptions/subid/resourcegroups/nad-core'. 
```
So the approach of trying to distinguish between the two case (not having permission to read a valid resource group _vs_ trying to read a non-existent resource group) appears to have hit a dead end.

**Comment by MyronFanQiu** [comment]
@jdashton I see. Looks like it will first do the authorization anyway. I will discuss with our members. Personally, I think it's reasonable to delete the check for the existence of resource group. We should not assume greater authorization than a caller may actually have. I hope your PR can be merged soon and will inform you with any update. Thank you very much~ 👍

**Comment by Fred-sun** [comment]
@MyronFanQiu Any update? Please help update when you're free! Thanks a lot!

**Comment by Fred-sun** [comment]
@MyronFanQiu

**Comment by Fred-sun** [comment]
@MyronFanQiu  Can you update the results of the discussion? Thank you very much!

**Comment by Fred-sun** [comment]
@MyronFanQiu @jdashton The relate PR has been merged, maybe this issue can be closed. Please check, thank you very much!

**Comment by jdashton** [comment]
Can we get a release, even a point release, of azure_preview_modules that includes this change? I reinstalled yesterday, and this fix was still not present.

**Comment by Fred-sun** [comment]
@jdashton Thanks for your trying. I check the repo, these changed already merged into [master branch](https://github.com/Azure/azure_preview_modules/blob/master/library/azure_rm_dnsrecordset.py), please check, Thank you very much!

**Comment by jdashton** [comment]
> @jdashton Thanks for your trying. I check the repo, these changed already merged into [master branch](https://github.com/Azure/azure_preview_modules/blob/master/library/azure_rm_dnsrecordset.py), please check, Thank you very much!

I agree that the commit and PR have been merged into master. Thank you! 

However, the instructions for this repo say to use this command to install the modules:

`ansible-galaxy install azure.azure_preview_modules`

Even when I use the `--force command`, I still get only version 2.8.0, which does not include the merged code. That's why I'm suggesting that you and your team create a new release, at least version 2.8.1, that has the newer code in it.

Thanks!
